contracts/CrossChain.sol

```diff
@@ -322,9 +398,141 @@ contract CrossChain is System, ICrossChain, IParamSubscriber{
+      uint16 cancelTransferApproveThreshold = BytesToTypes.bytesToUint16(32, value);
+      require(cancelTransferApproveThreshold > 0, "zero cancel transfer threshold");
+      approveThresholdMap[CANCEL_TRANSFER_PROPOSAL] = cancelTransferApproveThreshold;
     } else {
       require(false, "unknown param");
     }
     emit paramChange(key, value);
   }
-}
\ No newline at end of file
+
+  // BEP-171: Security Enhancement for Cross-Chain Module
+  function challenge(
+    // to avoid stack too deep error, using `uint64[4] calldata params`
+    // instead of  `uint64 height0, uint64 height1, uint64 packageSequence, uint8 channelId`
+    uint64[4] calldata params, // 0-height0, 1-height1, 2-packageSequence, 3-channelId,
+    bytes calldata payload0,
+    bytes calldata payload1,
+    bytes calldata proof0,
+    bytes calldata proof1
+  )
+  onlyInit
+  blockSynced(params[0])
+  blockSynced(params[1])
+  channelSupported(uint8(params[3]))
+  whenNotSuspended
+  external {
+    // the same key with different values (payloads)
+    require(keccak256(payload0) != keccak256(payload1), "same payload");
+
+    bytes memory _key;
+    uint64 _packageSequence;
+    uint8 _channelId;
+    {
+      _packageSequence = params[2];
+      _channelId = uint8(params[3]);
+      _key = generateKey(_packageSequence, _channelId);
+      bytes32 _keyHash = keccak256(_key);
+      require(!challenged[_keyHash], "already challenged");
+
+      // if succeeding in challenge
+      challenged[_keyHash] = true;
+    }
+
+    // verify payload0 + proof0
+    {
+      uint64 _height0 = params[0];
+      bytes memory _payload0 = payload0;
+      bytes memory _proof0 = proof0;
+      bytes32 _appHash0 = ILightClient(LIGHT_CLIENT_ADDR).getAppHash(_height0);
+      require(MerkleProof.validateMerkleProof(_appHash0, STORE_NAME, _key, _payload0, _proof0), "invalid merkle proof0");
+    }
+
+    // verify payload1 + proof1
+    {
+      uint64 _height1 = params[1];
+      bytes memory _payload1 = payload1;
+      bytes memory _proof1 = proof1;
+      bytes32 _appHash1 = ILightClient(LIGHT_CLIENT_ADDR).getAppHash(_height1);
+      require(MerkleProof.validateMerkleProof(_appHash1, STORE_NAME, _key, _payload1, _proof1), "invalid merkle proof1");
+    }
+
+    _emergencySuspend();
+    emit Challenge(msg.sender, _packageSequence, _channelId);
+  }
+
+  function emergencySuspend() onlyCabinet whenNotSuspended external {
+    bool isExecutable = _approveProposal(EMERGENCY_SUSPEND_PROPOSAL);
+    if (isExecutable) {
+      _emergencySuspend();
+    }
+  }
+
+  function reopen() onlyCabinet whenSuspended external {
+    bool isExecutable = _approveProposal(REOPEN_PROPOSAL);
+    if (isExecutable) {
+      isSuspended = false;
+      emit Reopen(msg.sender);
+    }
+  }
+
+  function cancelTransfer(address tokenAddr, address _attacker) onlyCabinet external {
+    bool isExecutable = _approveProposal(CANCEL_TRANSFER_PROPOSAL);
```

Comment:
the `tokenAddr` and `_attacker ` are only used in the last vote.

Reply:
rename `_attacker ` to `attacker`

Reply:
fixed